ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/Identifiers.scala

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api
+
+import com.daml.ledger._
+import com.daml.ledger.api.v1.transaction.{Transaction, TransactionTree}
+
+/**
+  * Extracts identifiers from Protobuf messages to correlate traces.
+  */
+object Identifiers {
+
+  /**
+    * Extract identifiers from a transaction message.
+    */
+  def fromTransaction(transaction: Transaction): Map[String, String] = {
+    val attributes = Map.newBuilder[String, String]
+    if (!transaction.offset.isEmpty) attributes += OffsetKey -> transaction.offset
```

Comment:
This could be slightly more readable as `transaction.offset.foreach(attributes += OffsetKey -> _)` etc.

Reply:
It looks like you might be assuming that `offset` is an `Option`, but it's a `String`. Not all of `offset`, `commandId`, etc. are always set, and I want to avoid adding empty attributes, hence the `if (!empty)` bit. `offset.foreach` would add each character, which is not what we want here.